### ISSUE TITLE
Workflow does not complete if `count_reads` is `true` but  `run_static_html` is `false`

### ISSUE THREAD

**Issue by mikej888** [riboviz/riboviz#416]
As `count_reads` can be `true` but `run_static_html` can be `false`, I wanted to see how the workflow would behave under this configuration (i.e. the user wants to count reads but doesn't want HTML output) . Currently the workflow terminates without invoking collateTpms or countReads.

Branch `develop`, commit 66ccfb5 Thu Aug 19 05:41:11 2021:

```console
$ rm -rf vignette/output/ vignette/index vignette/tmp/ work/
$ nextflow run prep_riboviz.nf -ansi-log false -params-file vignette/vignette_config.yaml --run_static_html false
...
[fb/9f30f9] Submitted process > generateStatsFigs (WTnone)
[2e/e0748a] Submitted process > generateStatsFigs (WT3AT)
Finished processing sample: WTnone
[e8/223a8a] Submitted process > renameTpms (WTnone)
Finished processing sample: WT3AT
No sample was visualised successfully

$ echo $?
1
$ ls vignette/output/
WT3AT  WTnone
```

The workflow has prematurely terminated and both `read_counts.tsv` and `TPMs_collated.tsv` are missing :-(

Branch `parallel-count-reads-386`, commit d1bffa2 Mon Aug 16 11:17:52 2021:

```console
$ rm -rf vignette/output/ vignette/index vignette/tmp/ work/
$ nextflow run prep_riboviz.nf -ansi-log false -params-file vignette/vignette_config.yaml --run_static_html false
...
[bd/9641ca] Submitted process > generateStatsFigs (WTnone)
[22/68012b] Submitted process > generateStatsFigs (WT3AT)
Finished processing sample: WTnone
[56/d05e19] Submitted process > renameTpms (WTnone)
Finished processing sample: WT3AT
No sample was visualised successfully

$ echo $?
1
$ ls vignette/output/
WT3AT  WTnone
```

Same issue.

In `develop`, the `countReads` process has a dependency on channel `collate_tpms_sample_ids` from `collateTpms`. This ensures that `countReads` doesn't start until `collateTpms` has complete and processing of samples (not counting `staticHTML`) has completed.

In `parallel-count-reads-386` (for #386), the `countReads` process dependency on channel `collate_tpms_sample_ids` from `collateTpms` was updated by @canuse to depend on `static_html_sample_ids` from the `staticHTML` process. This ensures that `countReads` doesn't start until processing of all samples in `staticHTML` has completed. This prevents these two compute-intensive tasks from happening concurrently.

In both cases the workflow needs to be studied to see how it can be redesigned to avoid this situation. On a more general point, the more optional steps that are introduced into the workflow the more likely such issues might arise.

**Comment by mikej888** [comment]
Though I found this via looking at #386, given it affects `develop` too I don't think it should block any future merge of #386's branch into `develop`. Nor is it critical - users just have to not use that configuration combination!

**Comment by mikej888** [comment]
`prep_riboviz.nf` has:

```
process staticHTML {
    ...
    output:
      val sample_id into finished_viz_sample_id
      file "${sample_id}_output_report.html" into static_html_html
    when:
      params.run_static_html
    ...

finished_viz_sample_id
    .ifEmpty { exit 1, "No sample was visualised successfully" }
    .view { "Finished visualising sample: ${it}" }
```

If `run_static_html` is `false` then `staticHTML` produces no output onto `finished_viz_sample_id` so it will always be empty and so `finished_viz_sample_id` will always immediately terminate the workflow. The exit needs to be made dependant, somehow, on `run_static_html` also.

**Comment by mikej888** [comment]
In addition, for the `parallel-count-reads-386` branch, how to ensure that `countReads` triggers even if `staticHTML` produces no output? 

OR

How to ensure that `countReads` and `staticHTML` do not execute concurrently? Is there another approach not involving input/output channels?

**Comment by mikej888** [comment]
Commit 30a73c8 in [workflow-bug-fix-416](https://github.com/riboviz/riboviz/tree/workflow-bug-fix-416) has a bug fix for `develop`:

```
// Create handler for finished_viz_sample_id channel, output by
// staticHTML, only if run_static_html is true i.e. if staticHTML
// executes.
if (params.run_static_html) {
  finished_viz_sample_id
      .ifEmpty { exit 1, "No sample was visualised successfully" }
      .view { "Finished visualising sample: ${it}" }
}
```


`run_static_html` `false`:
```
$ rm -rf vignette/output/ vignette/index vignette/tmp/ work/
$ nextflow run prep_riboviz.nf -ansi-log false -params-file vignette/vignette_config.yaml --run_static_html false
...
[2a/dd66ea] Submitted process > generateStatsFigs (WTnone)
[bd/b4c860] Submitted process > generateStatsFigs (WT3AT)
Finished processing sample: WTnone
[7d/e8102a] Submitted process > renameTpms (WTnone)
Finished processing sample: WT3AT
[bb/0fd8b3] Submitted process > renameTpms (WT3AT)
[20/d14aa7] Submitted process > collateTpms (WTnone, WT3AT)
[d2/64cfe9] Submitted process > countReads
Workflow finished! (OK)
$ echo $?
0
$ ls vignette/output/
read_counts.tsv  TPMs_collated.tsv  WT3AT  WTnone
```

`run_static_html` `true`:

```console
$ rm -rf vignette/output/ vignette/index vignette/tmp/ work/
$ nextflow run prep_riboviz.nf -ansi-log false -params-file vignette/vignette_config.yaml 
...
[c1/9f5791] Submitted process > generateStatsFigs (WTnone)
[58/6f4ebe] Submitted process > generateStatsFigs (WT3AT)
Finished processing sample: WTnone
[47/d34dec] Submitted process > staticHTML (WTnone)
[27/ae2aec] Submitted process > renameTpms (WTnone)
Finished processing sample: WT3AT
[0b/50a3c9] Submitted process > renameTpms (WT3AT)
[2f/34cc4f] Submitted process > staticHTML (WT3AT)
[a0/e31978] Submitted process > collateTpms (WTnone, WT3AT)
[8f/d3bc17] Submitted process > countReads
Finished visualising sample: WTnone
Finished visualising sample: WT3AT
Workflow finished! (OK)
$ echo $?
0
$ ls vignette/output/
read_counts.tsv  TPMs_collated.tsv  WT3AT  WTnone
```

**Comment by mikej888** [comment]
For `parallel-count-reads-386`, merging into `develop` will apply the above fix to prevent the workflow terminating prematurely if `run_static_html` is false.

At present `countReads` has the following inputs to forge it not to execute until previous processes have completed:

* `develop`: `collate_tpms_sample_ids` from `collateTpms`.
* `parallel-count-reads-386`, `static_html_sample_ids` from `staticHTML`

Perhaps combining these two approaches, and using a condition, would work so that `countReads` executes when `collateTpms` has completed if `run_static_html` is `false` or when `staticHTML` has completed if `run_static_html` is `true`.

```
if (params.run_static_html) {
  count_reads_sample_ids = static_html_sample_ids 
} else {
  count_reads_sample_ids = collate_tpms_sample_ids
}

process collateTpms {
    input:
        val sample_ids from count_reads_sample_ids.collect()
        ...
    output:
}
```

Whether this would work depends on whether the channel assignment would work.

**Comment by mikej888** [comment]
Foregoing works. commit 0522458 of [parallel-countread-386-workflow-bug-fix-416](https://github.com/riboviz/riboviz/tree/parallel-countread-386-workflow-bug-fix-416) from `parallel-countread-386` with branch `workflow-bug-fix-416` merged in has the updated version.

Tested as follows:

```
$ rm -rf vignette/output/ vignette/index vignette/tmp/ work/
$ nextflow run prep_riboviz.nf -ansi-log false -params-file vignette/vignette_config.yaml --run_static_html false --count_reads false
$ echo $?
$ ls vignette/output/
# Check for presence/absence of reads_counts.tsv
$ ls vignette/output/WTnone
# Check for presence/absenence of WTnone_output_report.html
```

Each of the 4 combinations of  `run_static_html true|false` and `count_reads true|false` was tested.

**Comment by mikej888** [comment]
Merged `update-file-names-411` (ac1ad88 27/08/21) into `workflow-bug-fix-416`.  `workflow-bug-fix-416` pull request #419 to be merged into `develop` post-merge of `update-file-names-411` pull request #422.
